### PR TITLE
feat: performance test for InfluxQL vs Flux (influxDB#18088)

### PR DESCRIPTION
**Pull request by DStrand1**
Opening this as a draft because I still have a few questions about if everything here is done correctly:
- I have this under the `metaquery` use-case, but think that testing with an `iot` data_gen would be better for the test, should this be under that use-case then?
- @wbaker85 and I spoke about running over random time windows, but any time I tried to do a `AllInterval.RandWindow()` time, it would cause a panic. Should I loop over 5 minute ranges until I hit `AllInterval.End` instead?
- Do these queries themselves (Flux, InfluxQL) seem to line up well with the original issue? https://github.com/influxdata/influxdb/issues/18088

Additionally, I'm unclear on how to actually run these tests for performance data outside of a CI environment, so I have not gathered any performance data yet

### CODE REVIEW

**Comment by wbaker85** [comment]
>     * I have this under the `metaquery` use-case, but think that testing with an `iot` data_gen would be better for the test, should this be under that use-case then?

Semantically this definitely isn't a metaquery, so putting it with the `iot` use-case makes sense to me if that's the data you are running it with. It would be interesting to see what (if any) differences in the perf comparisons this query has compared to the [existing IoT query](https://github.com/influxdata/influxdb-comparisons/blob/db2f91c56f7fc3752761a5f0a92e7b63525b6aa4/bulk_query_gen/influxdb/influx_iot_common.go#L65-L69). That one has an `aggregateWindow` in it as well.

>     * @wbaker85 and I spoke about running over random time windows, but any time I tried to do a `AllInterval.RandWindow()` time, it would cause a panic. Should I loop over 5 minute ranges until I hit `AllInterval.End` instead?

You should be able to do something like this:
```
	interval := d.AllInterval.RandWindow(5 * time.Minute)

	if d.language == InfluxQL {
		query = fmt.Sprintf(`SELECT mean("co2_level") as "mean_value" FROM "air_quality_room" WHERE time > %s AND time < %s AND "room_id"='4' GROUP BY time(5m) FILL(null)`,
			interval.StartString(),
			interval.EndString())
	} else {...}
```

...that runs for me. Does that cause a panic for you?

>     * Do these queries themselves (Flux, InfluxQL) seem to line up well with the original issue? [Performance difference between Flux and InfluxQL influxdb#18088](https://github.com/influxdata/influxdb/issues/18088)

It seems like you will want to allow for a longer range to be specified. That issue has a range of 7 days and 30 days. I would test and profile with a 7 day duration and a 30 day duration and see what it looks like. From the reported issue it looks like the flux performance is about 20x worse in both cases, so just using a fixed time window is probably fine, but you should test it out and make sure that the difference is linear over a range of time windows, and make sure you use a large enough window to clearly illustrate the difference between flux and influxql. Perhaps 5 minutes won't be enough to do that.

I'm assuming you aren't trying to benchmark the query for which flux was actually faster?

**Comment by DStrand1** [comment]
With these last commits, these queries will now work and run perfectly. Trying to recreate the queries demonstrated in the original issue, I got these values:
|                | 7 Day Raw | 7 Day (5m Window) | 30 Day (5m Window) |
|----------------|:---------:|:-----------------:|:------------------:|
| ORIGINAL       |           |                   |                    |
| InfluxQL (1.8) |    1.8s   |       46.3ms      |       201.3ms      |
| Flux (2.x)     |   26.7ms  |        1.3s       |        6.2s        |
|                |           |                   |                    |
| NEW            |           |                   |                    |
| InfluxQL (1.8) |   0.28ms  |       0.28ms      |        0.3ms       |
| Flux (2.x)     |  10.89ms  |      11.99ms      |       12.23ms      |

I'm not completely sold on these values, but to me this indicates two things:
- Flux is more performant on average for getting raw data of large datasets, where InfluxQL will be slower when the dataset is sufficiently large
- Flux is slower (still) on these types of queries. Doing some quick math on the ratios of these values:
    - 7 Day (5m Window) InfluxQL is ~28x more efficient in original test, and ~42x more efficient in current test
    - 30 Day (5m Window) InfluxQL is ~31x more efficient in original test, and ~41x more efficient in current test
    
While this data is not perfect, I think it still indicates that this issue *does* exist

File: bulk_query_gen/influxdb/influx_iot_aggregate_keep.go [hunk 1]
```diff
@@ -0,0 +1,32 @@
+package influxdb
+
+import (
+	bulkQuerygen "github.com/influxdata/influxdb-comparisons/bulk_query_gen"
+	"time"
+)
+
+// InfluxIotAggregateKeep produces metaqueries that will test performance
```
Comment (wbaker85) [inline bulk_query_gen/influxdb/influx_iot_aggregate_keep.go:8]:
```suggestion
// InfluxIotAggregateKeep produces queries that will test performance
```

File: bulk_query_gen/influxdb/influx_iot_common.go [hunk 2]
```diff
@@ -76,3 +76,28 @@ func (d *InfluxIot) averageTemperatureDayByHourNHomes(qi bulkQuerygen.Query, nHo
 	q := qi.(*bulkQuerygen.HTTPQuery)
 	d.getHttpQuery(humanLabel, interval.StartString(), query, q)
 }
+
+func (d *InfluxIot) IotAggregateKeep(qi bulkQuerygen.Query) {
+	interval := d.AllInterval.RandWindow(7 * 24 * time.Hour)
```
Comment (wbaker85) [inline bulk_query_gen/influxdb/influx_iot_common.go:81]: We might want to make this window size configurable via a flag at some point, but I think it's ok for now.

**Review by wbaker85 — APPROVED** [review]
LGTM, just one minor cleanup thing in a comment that I noted.

File: bulk_query_gen/influxdb/influx_iot_common.go [hunk 3]
```diff
@@ -76,3 +76,28 @@ func (d *InfluxIot) averageTemperatureDayByHourNHomes(qi bulkQuerygen.Query, nHo
...
+
+	var query string
+	if d.language == InfluxQL {
+		query = fmt.Sprintf(`SELECT mean("co2_level") as "mean_value" FROM "air_quality_room" WHERE time > '%s' AND time < '%s' AND "room_id"='4' GROUP BY time(5m) FILL(null)`,
+			interval.StartString(),
+			interval.EndString())
+	} else {
+		query = fmt.Sprintf(`from(bucket: "%s") `+
+			`|> range(start: %s, stop: %s) `+
+			`|> filter(fn: (r) => r._measurement == "co2_level" and r._field == "air_quality_room")`+
```
Comment (danxmoran) [inline bulk_query_gen/influxdb/influx_iot_common.go:91]: I think this is flipped: when I run the IoT data generation locally, I get a measurement of `air_quality_room` with a field of `co2_level`. I see the same in your InfluxQL query above has the same

**Comment by danxmoran** [comment]
I ran a pair of test queries locally w/ the predicate flipped, and saw the difference in performance is even more drastic than what's posted in the comments above. Based on the profiling output it looks like `keep` is preventing the window-aggregate pushdown.

**Review by danxmoran — APPROVED** [review]

**Pull request merged**